### PR TITLE
disable kv cache broadcast for better performance

### PR DESCRIPTION
**Pull request by nazanin-beheshti**
### Details:

 For MSFT Orca model, it has different num_head for query and key/value. It leads to a broadcast behavior in the graph which introduce a big overhead for the performance. This ticket is created to eliminate the broadcast and get good conformance for Orca pipeline.

stateless-before-after

### Tickets:
 - *CVS-186733*

### AI Assistance:
 - *AI assistance used: no*

### CODE REVIEW

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 1]
```diff
@@ -214,7 +214,7 @@ TEST(type_prop, scaled_dot_product_attention_mixed_shape_infer_5_inputs) {
     const auto op =
         std::make_shared<op::v13::ScaledDotProductAttention>(query, key, value, attention_mask, scale, causal);
     EXPECT_EQ(op->get_output_element_type(0), element::f64);
-    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{4, 3, {4, 5}, {3, 7}}));
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{{2, 4}, 3, {4, 5}, {3, 7}}));
```
Comment (Kotomi-Du) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp]: How broadcast behavior impacts the shape which has range?

**Comment by Kotomi-Du** [comment]
build_jenkins

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 2]
```diff
@@ -42,11 +42,7 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
     const auto& key_rank = key.rank();
     if (key_rank.is_static()) {
         const bool& key_input_correctness =
-            key_rank.get_length() >= 3 &&
-            TRShape::broadcast_merge_into(n_dims,
```
Comment (Kotomi-Du) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:46]: In your case, the rank of Q and K/V are same, only with different dimensions. To narrow down the impact in general, is it better to return success when ranks are same in the logic below? 

Image

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 3]
```diff
@@ -28,6 +28,19 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
+    const auto& key = input_shapes[1];
+    const auto& key_rank = key.rank();
+
+    size_t kv_num_heads_factor = 1;
+    if ((n_dims_rank.is_static() && key_rank.is_static()) && (n_dims_rank.get_length() == key_rank.get_length())) {
+        const auto& last_dim = *(n_dims.end() - 1);
+        const auto& key_dim =
+            (key_rank.is_static() && key_rank.get_length() >= 3) ? *(key.end() - 3) : ov::Dimension(1);
+        if (last_dim.is_static() && key_dim.is_static() && key_dim.get_length() != 0) {
+            kv_num_heads_factor = last_dim.get_length() / key_dim.get_length();
```
Comment (Kotomi-Du) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: @nazanin-beheshti  For case K shape [1,10,2048,128], Q shape [1,40,2048,128] and output shape is [1,40,2048,128], is kv_num_heads_factor = 128/10?

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 4]
```diff
@@ -354,6 +354,7 @@ TEST(type_prop, scaled_dot_product_unsupported_key_shape) {
         AssertFailure,
         testing::HasSubstr("Key input shape not compatible with other inputs."));
 }
+
```
Comment (Kotomi-Du) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp]: please remove unnecessary change

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 5]
```diff
@@ -364,7 +365,7 @@ TEST(type_prop, scaled_dot_product_unsupported_value_shape) {
     OV_EXPECT_THROW(
         auto op = std::make_shared<op::v13::ScaledDotProductAttention>(query, key, value, attention_mask, causal),
         AssertFailure,
-        testing::HasSubstr("Value input shape not compatible with other inputs."));
+        testing::HasSubstr("Attention mask input shape not compatible with other inputs."));
```
Comment (Kotomi-Du) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp]: it seems you add some unintentional change which is incorrect.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 2 above)
Comment (Kotomi-Du) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:46]: The name of last_dim is confusing, I thought it is the last dim of output shape. The same for  kv_num_head_factor. Please rename it for better understanding.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 6]
```diff
@@ -28,6 +28,21 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
     auto output_shapes = std::vector<TRShape>{input_shapes[0]};
     auto& n_dims = output_shapes[0];
     const auto& n_dims_rank = n_dims.rank();
+    const auto& key = input_shapes[1];
+    const auto& key_rank = key.rank();
+
+    size_t kv_num_heads_factor = 1;
+    if ((n_dims_rank.is_static() && key_rank.is_static()) && (n_dims_rank.get_length() == key_rank.get_length())) {
+        const ov::Dimension last_dim =
+            (n_dims_rank.is_static() && n_dims_rank.get_length() >= 3) ? *(n_dims.end() - 3) : ov::Dimension(1);
```
Comment (Kotomi-Du) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: the name of last_dim is confusing, I thought it is the last dim of output shape. Please correct. Suggested code
```
const ov::Dimension q_num_head = n_dims_rank.get_length() >= 3 ? *(n_dims.end() - 3) : ov::Dimension(1);
const ov::Dimension k_num_head = key_rank.get_length() >= 3 ? *(key.end() - 3) : ov::Dimension(1);   //static is checked in the condition.
if (kv_num_head != q_num_head) {
     skip_broadcast_merge = true;
}
```

**Review by Kotomi-Du — CHANGES_REQUESTED** [review]

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp (hunk 5 above)
Comment (nazanin-beheshti) [reply src/core/tests/type_prop/scaled_dot_product_attention.cpp]: Updated

**Review by Kotomi-Du — APPROVED** [review]

**Comment by CuriousPanCake** [comment]
Could you please add tests for the changes you made?

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 7]
```diff
@@ -28,6 +28,23 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
+        const ov::Dimension k_num_head =
+            (key_rank.is_static() && key_rank.get_length() >= 3) ? *(key.end() - 3) : ov::Dimension(1);
+
+        if (q_num_head.is_static() && k_num_head.is_static() && k_num_head != q_num_head) {
+            auto kv_num_heads_factor = q_num_head.get_length() / k_num_head.get_length();
+            if (kv_num_heads_factor > 1)
+                skip_broadcast_merge = true;
+        }
+    }
```
Comment (mmikolajcz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: According to spec, dimensions [N, ...] between Q/K/V should be numpy-broadcastable to each other and are not limited to any rank.
So, with this check it will use just on last element of this sub-shape in Q/K (V isn't used) to decide whether to skip whole broadcasting, even if other dims should be broadcasted or they are not compatible, breaking backward compatibility of SDPA.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 8]
```diff
@@ -28,6 +28,23 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
+
+    bool skip_broadcast_merge = false;
+    if ((n_dims_rank.is_static() && key_rank.is_static()) && (n_dims_rank.get_length() == key_rank.get_length())) {
+        const ov::Dimension q_num_head =
+            (n_dims_rank.is_static() && n_dims_rank.get_length() >= 3) ? *(n_dims.end() - 3) : ov::Dimension(1);
+        const ov::Dimension k_num_head =
+            (key_rank.is_static() && key_rank.get_length() >= 3) ? *(key.end() - 3) : ov::Dimension(1);
+
+        if (q_num_head.is_static() && k_num_head.is_static() && k_num_head != q_num_head) {
+            auto kv_num_heads_factor = q_num_head.get_length() / k_num_head.get_length();
```
Comment (mmikolajcz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: What if not Q heads are note equally divisible by K? Please update spec and add test cases

**Review by mmikolajcz — CHANGES_REQUESTED** [review]
In current form, this change breaks current SDPA spec contract https://docs.openvino.ai/nightly/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.html

- GQA special case isn't mentioned by spec
- Test cases should include test for GQA special broadcasting and divisibility validation
- skip_broadcast_merge breaks numpy-broadcasting rules for Q/K/V [N, ...] subshape

To work around those limitations, maybe SDPA should instead be extended by attribute with GQA mode or internal GQA op used in transformation should be published?

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 9]
```diff
@@ -57,17 +43,18 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
-            key_input_correctness = key_rank.get_length() >= 3 && DimType::merge(e_dim, e_dim, *(key.end() - 1));
-        } else {
-            key_input_correctness =
-                key_rank.get_length() >= 3 &&
-                TRShape::broadcast_merge_into(n_dims,
-                                              TRShape(std::vector<DimType>(key.begin(), key.end() - 2)),
-                                              AutoBroadcastType::NUMPY) &&
-                DimType::merge(e_dim, e_dim, *(key.end() - 1));
+        if (n_dims.rank().get_length() >= 3 && key_rank.get_length() >= 3) {
+            const ov::Dimension q_num_head = *(n_dims.end() - 1);
```
Comment (Kotomi-Du) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:48]: the name is incorrect, the last dimension of query shape is head dimension, not num of head.
Also please explain why q_head_dimension%key_num_head !=0 could decide if it is gqa mode.
The core idea of GQA is many query heads share fewer K/V heads, so q_head_num should be always greater than kv_head_num.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 7 above)
Comment (nazanin-beheshti) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: I try to extend SDPA with GQA mode attribute, but it needs a lot of changes in many files.
With latest changes, broadcast_merge_into is not completely skipped. The batch dimension is still broadcast if we have the q_head > k_head and only head dimension broadcast is skipped.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 9 above)
Comment (nazanin-beheshti) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:48]: I change the condition that if query head size is greater than Key head size, we set the gqa_mode variable and skip head_dimension broadcast, but batch broadcast is still there.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 10]
```diff
@@ -38,31 +43,67 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
-        const bool& key_input_correctness =
-            key_rank.get_length() >= 3 &&
-            TRShape::broadcast_merge_into(n_dims,
-                                          TRShape(std::vector<DimType>(key.begin(), key.end() - 2)),
-                                          AutoBroadcastType::NUMPY) &&
-            DimType::merge(e_dim, e_dim, *(key.end() - 1));
+        if (key_rank.get_length() >= 3) {
+            const ov::Dimension q_num_head = *(n_dims.end() - 1);
+            const ov::Dimension k_num_head = (key_rank.is_static()) ? *(key.end() - 3) : ov::Dimension(1);
+            if (q_num_head.is_static() && k_num_head.is_static()) {
```
Comment (Kotomi-Du) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:50]: This is a generic logic for both value and key, not good place to put. 
It seems you are trying to keep batch dimension for broadcasting, please refactor the code to be more readable.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 11]
```diff
@@ -36,33 +42,60 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
         l_dim = *(n_dims.end() - 2);
         e_dim = *(n_dims.end() - 1);
         n_dims.resize(n_dims.size() - 2);
+        n_dims_sub = TRShape(std::vector<DimType>(n_dims.begin(), n_dims.end() - 1));
+        if ((key_rank.is_static() && key_rank.get_length() >= 3) && value_rank.is_static() &&
+            value_rank.get_length() >= 3) {
+            const ov::Dimension q_num_head = *(n_dims.end() - 1);
+            const ov::Dimension k_num_head = (key_rank.is_static()) ? *(key.end() - 3) : ov::Dimension(1);
```
Comment (mmikolajcz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:49]: Check whether key rank is static is unnecessary and could lead to confusion. IF before already secures that it has to be static, and if rank would be dynamic, assuming it has Dimension(1) is risky since this would always mean GQA

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 12]
```diff
@@ -36,33 +42,60 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
+            if (q_num_head.is_static() && k_num_head.is_static()) {
+                if ((q_num_head.get_length() % k_num_head.get_length() == 0) &&
+                    (q_num_head.get_length() != k_num_head.get_length())) {
+                    gqa_mode = true;
```
Comment (mmikolajcz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:53]: To workaround backward compatibility, GQA mode shouldn't be inferred from shape_infer, otherwise this would be an undocumented hidden mode in public operator.
In current form it checks only for last batch dim (head) for keys, leading to completly ignoring equivalent head dim for value. If value head would be specified, key head per spec and previous behavior should be broadcasted. **This leads to backward incompatible behavior that would ignore head dimension in value.**

To be certain, tests for plugins (CPU/GPU/NPU/TEMPLATE) should be extended within PR scope to ensure that this change won't lead to unexpected behaviors.

Generally changed functionality requires new op version, however sometimes addition to previous op are accepted as long as default behavior matches original version or previous behavior was clearly undefined/unsupported. In current form, safest solution would probably be adding attribute for GQA like in torch that forces additional assertions: https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.scaled_dot_product_attention.html
You can refer to @mitruska extension of SDPA with new input for reference https://github.com/openvinotoolkit/openvino/pull/31786. But it is usually exception that needs to be discussed.

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 13]
```diff
@@ -537,3 +537,14 @@ TEST(type_prop, scaled_dot_product_attention_sink_input_wrong_last_dim) {
                     AssertFailure,
                     testing::HasSubstr("Sink input has not compatible shape."));
 }
+
+TEST(type_prop, scaled_dot_product_attention_static_gqa_q_heads_greater_than_kv_heads) {
+    const auto query = std::make_shared<op::v0::Parameter>(element::f32, PartialShape{1, 40, 128, 128});
+    const auto key = std::make_shared<op::v0::Parameter>(element::f32, PartialShape{1, 10, 128, 128});
+    const auto value = std::make_shared<op::v0::Parameter>(element::f32, PartialShape{1, 10, 128, 128});
```
Comment (mmikolajcz) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp:544]: Shape tests should also test assertions for unsupported shapes. For example, value of shape {1, 3, 128, 128}, {1, 1, 128, 128}, etc would still pass even if it is invalid or wouldn't be broadcasted.
